File: apps/worker/src/progress.ts
import { useLocalhostForDockerHost } from "./urlUtils";

export interface ProgressReportPayload {
  videoId: string;
  organizationId?: string;
  status: "PROCESSING" | "READY" | "FAILED";
  progress: number;
  [key: string]: any;
}

export class ProgressReporter {
  private videoId: string;
  private organizationId: string;
  private callbackUrl?: string;
  private lastReportedProgress: number = -1;
  private lastReportedTime: number = 0;
  private minProgressDelta: number = 5;
  private minTimeDeltaMs: number = 5 * 60 * 1000; // 5 minutes
  private isSending: boolean = false;

  constructor(videoId: string, organizationId: string = "default", callbackUrl?: string) {
    this.videoId = videoId;
    this.organizationId = organizationId;
    this.callbackUrl = callbackUrl;
  }

  /**
   * Reports progress if forced or if progress has increased by >= 5% or 5 minutes have elapsed.
   */
  public async report(
    currentProgress: number,
    status: "PROCESSING" | "READY" | "FAILED" = "PROCESSING",
    force: boolean = false,
    extraData: Record<string, any> = {}
  ): Promise<void> {
    const progress = Math.min(100, Math.max(0, Math.floor(currentProgress)));
    const now = Date.now();

    const progressDelta = Math.abs(progress - this.lastReportedProgress);
    const timeDelta = now - this.lastReportedTime;

    const shouldReport =
      force ||
      this.lastReportedProgress === -1 ||
      progress === 100 ||
      status !== "PROCESSING" ||
      progressDelta >= this.minProgressDelta ||
      timeDelta >= this.minTimeDeltaMs;

    if (!shouldReport) return;
    if (this.isSending && !force) return;

    this.lastReportedProgress = progress;
    this.lastReportedTime = now;

    if (!this.callbackUrl) {
      console.log(`[ProgressReporter] videoId: ${this.videoId} progress: ${progress}% (No callbackUrl)`);
      return;
    }

    const rawPayload: ProgressReportPayload = {
      videoId: this.videoId,
      organizationId: this.organizationId,
      status,
      progress,
      ...extraData,
    };

    const payload = useLocalhostForDockerHost(rawPayload);
    console.log(`[ProgressReporter] Reporting progress ${progress}% for video ${this.videoId} (${status})`);

    const workerSecret = process.env.WORKER_SECRET_TOKEN;
    const headers: Record<string, string> = { "Content-Type": "application/json" };
    if (workerSecret) {
      headers["Authorization"] = `Bearer ${workerSecret}`;
      headers["x-worker-secret"] = workerSecret;
    }

    this.isSending = true;
    try {
      const res = await fetch(this.callbackUrl, {
        method: "POST",
        headers,
        body: JSON.stringify(payload),
      });

      if (!res.ok) {
        const errorText = await res.text().catch(() => "");
        console.error(`[ProgressReporter Error] HTTP ${res.status} from ${this.callbackUrl}: ${errorText}`);
      } else {
        console.log(`[ProgressReporter] Successfully posted ${progress}% callback to ${this.callbackUrl}`);
      }
    } catch (err: any) {
      console.error(`[ProgressReporter Error] Failed to post callback to ${this.callbackUrl}:`, err?.message || err);
    } finally {
      this.isSending = false;
    }
  }
}

/**
 * Calculates overall transcoding progress.
 * Transcoding total weight = 80%.
 * If there are N renditions, each completed rendition accounts for (80 / N)%.
 * Intra-rendition progress (0-100%) scales the current rendition's share.
 */
export function calculateTranscodeProgress(
  renditionIndex: number,
  totalRenditions: number,
  renditionPercent: number
): number {
  if (totalRenditions <= 0) return 80;
  const clampedRenditionPercent = Math.min(100, Math.max(0, renditionPercent));
  const completedWeight = (renditionIndex / totalRenditions) * 80;
  const currentWeight = (clampedRenditionPercent / 100) * (80 / totalRenditions);
  return Math.min(80, completedWeight + currentWeight);
}

/**
 * Calculates upload phase progress.
 * Upload phase spans from 80% to 100% (weight = 20%).
 */
export function calculateUploadProgress(uploadRatio: number): number {
  const clampedRatio = Math.min(1, Math.max(0, uploadRatio));
  return 80 + clampedRatio * 20;
}

/**
 * Utility to parse ffmpeg timemark string "HH:MM:SS.ms" into seconds.
 */
export function parseTimemarkToSeconds(timemark: string | undefined): number {
  if (!timemark) return 0;
  const parts = timemark.split(":");
  if (parts.length === 3) {
    return parseFloat(parts[0]) * 3600 + parseFloat(parts[1]) * 60 + parseFloat(parts[2]);
  } else if (parts.length === 2) {
    return parseFloat(parts[0]) * 60 + parseFloat(parts[1]);
  }
  return parseFloat(timemark) || 0;
}

File: apps/worker/go/progress.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

type ProgressReporter struct {
	videoId              string
	organizationId       string
	callbackUrl          string
	lastReportedProgress int
	lastReportedTime     time.Time
	minProgressDelta     int
	minTimeDelta         time.Duration
	isSending            bool
	mu                   sync.Mutex
}

func NewProgressReporter(videoId, organizationId, callbackUrl string) *ProgressReporter {
	if organizationId == "" {
		organizationId = "default"
	}
	return &ProgressReporter{
		videoId:              videoId,
		organizationId:       organizationId,
		callbackUrl:          callbackUrl,
		lastReportedProgress: -1,
		minProgressDelta:     5,
		minTimeDelta:         5 * time.Minute,
	}
}

func (pr *ProgressReporter) Report(currentProgress float64, status string, force bool, extraData map[string]interface{}) {
	pr.mu.Lock()

	if status == "" {
		status = "PROCESSING"
	}

	progress := int(math.Min(100, math.Max(0, math.Floor(currentProgress))))
	now := time.Now()

	progressDelta := int(math.Abs(float64(progress - pr.lastReportedProgress)))
	timeDelta := now.Sub(pr.lastReportedTime)

	shouldReport := force ||
		pr.lastReportedProgress == -1 ||
		progress == 100 ||
		status != "PROCESSING" ||
		progressDelta >= pr.minProgressDelta ||
		timeDelta >= pr.minTimeDelta

	if !shouldReport {
		pr.mu.Unlock()
		return
	}
	if pr.isSending && !force {
		pr.mu.Unlock()
		return
	}

	pr.lastReportedProgress = progress
	pr.lastReportedTime = now
	pr.isSending = true
	pr.mu.Unlock()

	defer func() {
		pr.mu.Lock()
		pr.isSending = false
		pr.mu.Unlock()
	}()

	if pr.callbackUrl == "" {
		fmt.Printf("[ProgressReporter] videoId: %s progress: %d%% (No callbackUrl)\n", pr.videoId, progress)
		return
	}

	payload := map[string]interface{}{
		"videoId":        pr.videoId,
		"organizationId": pr.organizationId,
		"status":         status,
		"progress":       progress,
	}
	if extraData != nil {
		for k, v := range extraData {
			payload[k] = v
		}
	}

	convertedPayload := UseLocalhostForDockerHost(payload)
	targetCallbackUrl := ReplaceLocalhost(pr.callbackUrl)

	fmt.Printf("[ProgressReporter] Reporting progress %d%% for video %s (%s)\n", progress, pr.videoId, status)

	bodyBytes, err := json.Marshal(convertedPayload)
	if err != nil {
		fmt.Printf("[ProgressReporter Error] Failed to marshal JSON payload: %v\n", err)
		return
	}

	req, err := http.NewRequest("POST", targetCallbackUrl, bytes.NewBuffer(bodyBytes))
	if err != nil {
		fmt.Printf("[ProgressReporter Error] Failed to create HTTP request: %v\n", err)
		return
	}

	req.Header.Set("Content-Type", "application/json")

	workerSecret := getEnvString(os.Getenv("WORKER_SECRET_TOKEN"))
	if workerSecret != "" {
		req.Header.Set("Authorization", "Bearer "+workerSecret)
		req.Header.Set("x-worker-secret", workerSecret)
	}

	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Printf("[ProgressReporter Error] Failed to post callback to %s: %v\n", targetCallbackUrl, err)
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		errorText, _ := io.ReadAll(resp.Body)
		fmt.Printf("[ProgressReporter Error] HTTP %d from %s: %s\n", resp.StatusCode, targetCallbackUrl, string(errorText))
	} else {
		fmt.Printf("[ProgressReporter] Successfully posted %d%% callback to %s\n", progress, targetCallbackUrl)
	}
}

func CalculateTranscodeProgress(renditionIndex, totalRenditions int, renditionPercent float64) float64 {
	if totalRenditions <= 0 {
		return 80.0
	}
	clampedPercent := math.Min(100.0, math.Max(0.0, renditionPercent))
	completedWeight := (float64(renditionIndex) / float64(totalRenditions)) * 80.0
	currentWeight := (clampedPercent / 100.0) * (80.0 / float64(totalRenditions))
	return math.Min(80.0, completedWeight+currentWeight)
}

func CalculateUploadProgress(uploadRatio float64) float64 {
	clampedRatio := math.Min(1.0, math.Max(0.0, uploadRatio))
	return 80.0 + (clampedRatio * 20.0)
}

func ParseTimemarkToSeconds(timemark string) float64 {
	if timemark == "" {
		return 0
	}
	parts := strings.Split(timemark, ":")
	if len(parts) == 3 {
		h, _ := strconv.ParseFloat(parts[0], 64)
		m, _ := strconv.ParseFloat(parts[1], 64)
		s, _ := strconv.ParseFloat(parts[2], 64)
		return h*3600 + m*60 + s
	} else if len(parts) == 2 {
		m, _ := strconv.ParseFloat(parts[0], 64)
		s, _ := strconv.ParseFloat(parts[1], 64)
		return m*60 + s
	}
	s, _ := strconv.ParseFloat(timemark, 64)
	return s
}
